pipeline/transform.py

```diff
@@ -1,23 +1,119 @@
+        return False
+    return MIN_CDI <= cdi <= MAX_CDI
+
+
+def clean_data(earthquake_data: list[dict]) -> list[dict]:
     '''
-    Function to clean the data extracted
-    Replaces any None values with actual values 
+    Function to clean the data extracted from the earthquake API.
+    - Converts 'felt' and 'cdi' columns to numeric values, replacing invalid or missing values with defaults.
+    - Replaces missing values in the 'alert' column with 'green' as the default value.
+    - Gets rid of milliseconds in timestamp.
     '''
+    logging.basicConfig(level=logging.INFO,
+                        format='%(asctime)s - %(levelname)s - %(message)s')
+
+    if not isinstance(earthquake_data, list):
+        logging.error("Expected list for earthquake_data, got %s",
+                      type(earthquake_data))
+        return []
+
+    logging.info("Cleaning earthquake data...")
     earthquake_df = pd.DataFrame(earthquake_data)
+
     earthquake_df["felt"] = pd.to_numeric(
-        earthquake_df["felt"], errors="coerce").fillna(0)
-    earthquake_df["alert"] = earthquake_df["alert"].fillna("green")
+        earthquake_df["felt"], errors="coerce").fillna(ZERO)
+    earthquake_df["alert"] = earthquake_df["alert"].fillna(GREEN)
     earthquake_df["cdi"] = pd.to_numeric(
-        earthquake_df["cdi"], errors="coerce").fillna(-1)
-    return earthquake_df
+        earthquake_df["cdi"], errors="coerce").fillna(MIN_CDI)
+
+    earthquake_df = validate_data(earthquake_df)
+
+    earthquake_df["at"] = earthquake_df['at'].apply(
+        lambda x: x.replace(microsecond=ZERO))
+
+    logging.info("Data cleaning complete.")
+    return earthquake_df.to_dict(orient='records')
+
+
+def validate_data(earthquake_df: pd.DataFrame) -> pd.DataFrame:
+    """Ensures that long, lat, magnitude, cdi readings are within a valid range"""
+    logging.info("Validating data...")
+
+    earthquake_df["latitude_valid"] = earthquake_df["latitude"].apply(
+        is_valid_latitude)
+    earthquake_df["longitude_valid"] = earthquake_df["longitude"].apply(
+        is_valid_longitude)
+    earthquake_df["magnitude_valid"] = earthquake_df["magnitude"].apply(
+        is_valid_magnitude)
+    earthquake_df["cdi_valid"] = earthquake_df["cdi"].apply(is_valid_cdi)
+
+    valid_earthquake_df = earthquake_df[earthquake_df["latitude_valid"] &
+                                        earthquake_df["longitude_valid"] &
+                                        earthquake_df["magnitude_valid"] &
+                                        earthquake_df["cdi_valid"]]
+
+    invalid_count = len(earthquake_df) - len(valid_earthquake_df)
+    if invalid_count > 0:
+        logging.warning(
+            "Removed %s invalid earthquake records.", invalid_count)
+
+    columns_to_exclude = ["latitude_valid",
+                          "longitude_valid", "magnitude_valid", "cdi_valid"]
+    valid_earthquake_df.drop(columns=columns_to_exclude, inplace=True)
```

Comment:
I have heard that inplace=True is discouraged as it may be deprecated in future. Thinking long-term, is there another way to handle this?

Reply:
I have updated it now.